Add fit model to CurveAnalysis

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

This PR is the first step to introduce group in #715. The purpose of the PR is to add a fit model class to `CurveAnalysis` and to remove `curve_fitter` from analysis option.

### Details and comments

According to the review comments from @yaelbh and @eggerdj, the complexity of the `CurveAnalysis` is likely an issue to further extend the class. This PR introduces new class `FitModel` and its subclass `SingleFitFunction` and `CompositeFitFunction` that user can also check to grasp the model without reading the code. For example:

```python
>>> from qiskit_experiments.curve_analysis.standard_analysis import ErrorAmplificationAnalysis
>>> ErrorAmplificationAnalysis().fit_model

SingleFitFunction(x, amp, d_theta, phase_offset, base, angle_per_gate)

>>> from qiskit_experiments.library.characterization.analysis import FineAmplitudeAnalysis
>>> FineAmplitudeAnalysis().fit_model

CompositeFitFunction(x, amp, d_theta, base; @ Fixed angle_per_gate, phase_offset)
```

Now you can recall `FineAmplitudeAnalysis` does  SPAM correction with extra experiment, and also some parameters are fixed during the fitting.

From software point of view, the complexity of multi objective optimization is all offloaded to the fit model, and the rest of code doesn't need to take care how the fit model is implemented. If there is only single curve in the analysis, it implicitly uses `SingleFitFunction` which implements lightweight logic to compute y data.

yaelbh commented:
@nkanazawa1989 I need more documentation in order to review this PR. Required are both inline comments inside methods, to describe the inner variables, their data structures and the manipulations that they're going through; and more detailed doc strings, with more explanation and examples.

Specifically I started with the new file `fit_model.py`. It looks like, when `fit_models` is not `None`, then its length should be equal to the number of fit functions. This is what I understand, but I'm not sure if I understand correctly, because such a requirement is not documented nor checked. I also understand that `fixed_parameters`, although it is a list like all the other constructor arguments, is not restricted to any length, because (again, to my understanding) it's a list of parameter names that are shared among all the fit functions. If so, then it should be documented, otherwise one may tend to think it's again (like the previous arguments) a list where each element corresponds to a fit function.

Trying to figure out what's the role of the `_uniton_params` data member, I thought it'd be easier if I first review the sub-class `SingleFitFunction`. But then in the `args` documentation, I don't understand what you mean by "Composite X values array" and "Variadic argument provided from the fitter". Also, the types are missing (I only know that `x` is an array, but I don't know the type of elements of the array).

When strings are involved - in signatures and fit models - it's important to specify the expected syntax of the strings. Namely, how to write a string that implements to a certain signature.

nkanazawa1989 commented:
I added more documentation. Basically this PR combines parameter mapping code distributed across multiple functions/files into a single place as `FitModel` so there is nothing changed drastically (i.e. this is just a cleanup). The `FitModel` is mainly called internally, thus I didn't provide enough description for the data model (remain unchanged as well). Anyways I hope new documentation added in https://github.com/Qiskit/qiskit-experiments/pull/726/commits/8be5f0c28e90272ad1129c36680bf6370290b33a helps your review.

These are the logic currently implementing the curve analysis:
https://github.com/Qiskit/qiskit-experiments/blob/d27fe41cec995072fa1db02a14d8437028d81a4e/qiskit_experiments/curve_analysis/curve_fit.py#L27-L35

and parameter mapping is done in
https://github.com/Qiskit/qiskit-experiments/blob/d27fe41cec995072fa1db02a14d8437028d81a4e/qiskit_experiments/curve_analysis/curve_fit.py#L326-L332

I also removed _Composite_ to avoid confusion. For the variadic arguments you can refer to the API docs of scipy curve fit. There is no detailed docs there too.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html

eggerdj commented on qiskit_experiments/curve_analysis/curve_data.py:
```diff
@@ -47,6 +48,20 @@ class SeriesDef:
+    signature: List[str] = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        """Implicitly parse fit function signature for fit function."""
```
```suggestion
        """Parse the fit function signature to extract the names of the variables."""
```

eggerdj commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -825,13 +935,24 @@ def _run_analysis(
+        # Prepare fit model
+        prepared_fit_func = self.fit_model
+
+        if self.__fixed_parameters__:
+            fixed_params = {p: self.options.get(p) for p in self.__fixed_parameters__}
+            prepared_fit_func.bind_parameters(**fixed_params)
+        else:
+            fixed_params = None
```
If we move the fixed parameters to default options and initialize the `fit_model` with them we can get ride of this logic.

eggerdj reviewed (changes requested):
I think we should use this PR as an opportunity to refactor how fixed_parameters are defined. Overall the class `FitModel` and its sub-classes seems like a nice idea.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,199 @@
+    thus it behaves like a Python function that the SciPy curve_fit solver accepts.
+    Note that the fit function there only accepts variadic arguments.
+
+    This class ties together the fit function and associated parameter names to
+    perform correct parameter mapping among multiple objective functions with different signatures,
+    in which some parameters may be excluded from the fitting when they are fixed.
```
Does the example I added in https://github.com/Qiskit/qiskit-experiments/pull/726/commits/8be5f0c28e90272ad1129c36680bf6370290b33a resolve this?

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,199 @@
+            for parameter in signature:
+                if parameter not in union_params and parameter not in fixed_parameters:
+                    union_params.append(parameter)
+        self._uniton_params = union_params
```
This will be breaking API change because @Parameters entry in the curve analysis returns a list of fit parameter values in this order. Of course there is another list for parameter name so user can always convert it into dict with `dict(zip(param_names, popts))`. However some external user may have already built own package to track this vector for monitoring, for example.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -236,6 +239,39 @@ class AnalysisExample(CurveAnalysis):
     #: List[str]: Fixed parameter in fit function. Value should be set to the analysis options.
     __fixed_parameters__ = list()
```
This is nice idea. Likely this is major change so do you think it should be in the separate PR?

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,265 @@
+    def __call__(self, x: np.ndarray, *params) -> np.ndarray:
+        pass
+
+    def bind_parameters(self, **kwparams):
+        """Assign values to the fixed parameters."""
+        bind_dict = {k: kwparams[k] for k in self._fixed_params.keys() if k in kwparams}
+        self._fixed_params.update(bind_dict)
```
This is still needed because user may update it via the set option after instance is initialized. If you don't want to allow this, fixed parameter dict should be in the class attribute rather than the default options.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -236,6 +239,39 @@ class AnalysisExample(CurveAnalysis):
+    # Automatically generated fitting functions of child class
+    _fit_model = None
+
+    def __init_subclass__(cls, **kwargs):
+        """Parse series definition of subclass and set fit function and signature."""
+
+        super().__init_subclass__(**kwargs)
```
This is not good direction in several reasons. The primary reason is simply the performance. Inspecting a callable (note that `inspect` is not lightweight computation) every time the class is instantiated will induce significant computation overhead (current code is kind of doing this) though this doesn't change from instance to instance. For example, if we run parallel experiment on some 100Q device, I guess this overhead cannot be ignored. Another reason is parsed function signature is used to generate a template of `p0` and `bounds` in the default options, which is class method thus the signature should be parsed before the class is instantiated.

For DRAG and Ham tomo, dynamic generation of fit model doesn't help because we also need filter kwargs. I don't want to tie filter kwargs to the fit model since this will induce unnecessary coupling between fitting and data processing, which we are trying to separate.

yaelbh replied:
These considerations must be documented somewhere (in a visible place) in the code. Can be at the beginning of `__init_subclass__`. Otherwise someone will come and say "hey, it's not supposed to be a class method, let's set this up at init time`, without being aware of the potential effect on large parallel experiments.

nkanazawa1989 replied:
Good point :)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -249,51 +285,135 @@ def __init__(self):
+    def fit_model(self) -> FitModel:
+        """Return a fit model for this analysis instance."""
+        # This should return a copy of instance.
+        # Note that fit model is class attribute though parameters can be bound.
```
This is already instance attribute but I don't think we should change the `_fit_model` to the instance variable. See my comment above.

yaelbh commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,199 @@
+        fixed_parameters: Optional[List[str]] = None,
+    ):
+        super().__init__(fit_functions, signatures, fit_models, fixed_parameters)
+        self.data_allocation = None
```
@nkanazawa1989 Then please add a comment "`self.data_allocation` is set by `CurveAnalysis`".

yaelbh replied:
I'm also not clear about the flow. It seems too tight to `CurveAnalysis`. In the constructor we set `self.data_allocation` to `None`, but then inside `__call__` we assume that it's already initialized (in the line `inds = self.data_allocation == i`). What happens in-between?

nkanazawa1989 replied:
One must set `self.data_allocation`. I've already written this in the class documentation 

> It seems too tight to CurveAnalysis

This is not quite strange as you said. For example, `AnalysisResultData` class defines `.chisq` value which defaults to `None`. User can set this value whenever they want, but should be populated before the data is uploaded to server. Then what is the object this class is tied to? Same for the fit model. 

Note that fit models are not exposed to users in the standard use case, so careful user guide is not necessary here. Probably adding explicit setter will make the code more readable, but from users, the setter for the `data_allocation` and just public member like now doesn't look differently.

nkanazawa1989 replied:
I added more documentation in https://github.com/Qiskit/qiskit-experiments/pull/726/commits/c327fa5f58c74f0017964b31e01828d9d4b9d4d4

Note that I did NOT change any data representation from the original implementation, and there is only single line documentation like this. I can maintain the code now.
https://github.com/Qiskit/qiskit-experiments/blob/9338e71bea9320a5b3a247a148ca137c35b76e46/qiskit_experiments/curve_analysis/curve_fit.py#L184-L195

yaelbh commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,257 @@
+
+        y = np.zeros(x.size, dtype=float)
+        for i, (func, sig) in enumerate(zip(self._fit_functions, self._signatures)):
+            inds = self.data_allocation == i
```
Maybe verify that `self.data_allocation` is not `None` anymore?

nkanazawa1989 replied:
This is called multiple times from scipy fitter. Adding redundant branching logic may degrade performance (I didn't check though)

nkanazawa1989 replied:
Note that usually this model is not called by end-users or custom curve analysis. But the logic calling the model exist in the base class of curve analysis. This means too much redundancy for readability or edge case will just degrade performance.

nkanazawa1989 replied:
Note that this is boolean index array. `data_allocation` is not a dict.
```
>>> vec = np.array([0, 0, 0, 0, 1, 0, 1, 1, 1])
>>> vec[1] 
0
>>> vec == 1
array([False, False, False, False,  True, False,  True,  True,  True])
```

yaelbh commented on releasenotes/notes/add-fit-model-and-remove-curve_fitter-arg-4a0effb5f9b88ba9.yaml:
```diff
@@ -0,0 +1,24 @@
+    added to the curve analysis module. These classes abstract the curve definition
+    hard-coded in the class attribute :attr:`CurveAnalysis.__series__` as a
+    Python function-like object. This may help user to grasp the fitting model
+    without reading the code.
```
I don't understand this paragraph (except for the first sentence). Please rephrase it.

yaelbh commented on releasenotes/notes/add-fit-model-and-remove-curve_fitter-arg-4a0effb5f9b88ba9.yaml:
```diff
@@ -0,0 +1,24 @@
+    without reading the code.
+upgrade:
+  - |
+    Analysis option `curve_fitter` of the :class:`CurveAnalysis` has be removed
```
```suggestion
    Analysis option `curve_fitter` of the :class:`CurveAnalysis` has been removed
```

yaelbh commented on releasenotes/notes/add-fit-model-and-remove-curve_fitter-arg-4a0effb5f9b88ba9.yaml:
```diff
@@ -0,0 +1,24 @@
+upgrade:
+  - |
+    Analysis option `curve_fitter` of the :class:`CurveAnalysis` has be removed
+    because of the serialization problem. To use custom curve fitting library,
```
Which serialization problem?

nkanazawa1989 replied:
You know, the callable is not serializable.

nkanazawa1989 replied:
This is not quite true. Please check the update
https://github.com/Qiskit/qiskit-experiments/pull/726/commits/cd19b811ad5a98d8d4fe7549ffa1ad6b4daa9df5

yaelbh commented on releasenotes/notes/add-fit-model-and-remove-curve_fitter-arg-4a0effb5f9b88ba9.yaml:
```diff
@@ -0,0 +1,24 @@
+developer:
+  - |
+    :meth:`CurveAnalysis.curve_fit` has been added to the curve analysis and
+    its subclasses. Now you can directly access to the core fitting code
```
```suggestion
    its subclasses. Now you can directly access the core fitting code
```

yaelbh commented on releasenotes/notes/add-fit-model-and-remove-curve_fitter-arg-4a0effb5f9b88ba9.yaml:
```diff
@@ -0,0 +1,24 @@
+    :meth:`CurveAnalysis.curve_fit` has been added to the curve analysis and
+    its subclasses. Now you can directly access to the core fitting code
+    with bare numpy arrays representing data to be fit.
+    This may help debugging of new fit function.
```
```suggestion
    This may help debugging of new fit functions.
```

yaelbh replied:
Please add a reference to an example that demonstrates how to use the new feature for debugging fit functions.

nkanazawa1989 replied:
There is no example, but user can directly input ndarrays without constructing `ExperimentData` container. Since there is no user guide how to construct experiment data, it is very tough to test fit functions in the curve analysis without reading the code of experiment data which is super long. I can remove
> This may help debugging of new fit functions.

If I need to write mode docs just for it.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,265 @@
+
+    .. math::
+
+        F(x, \Theta) = f_0(x_0, \vec{p}_0) \oplus f_1(x_1, \vec{p}_1) \oplus ...
```
It's already in the form of composition of functions.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -372,6 +499,27 @@ def _default_options(cls) -> Options:
+                "Option curve_fitter has been removed. Please directly override curve_fit method."
+            )
+
+        if "fixed_parameters" in fields:
```
No. With the previous #734 `fixed_parameters` became an instance options. This means, the parameter is no longer a part of the static model but this is something user can update after instance is created. When user update the option, this should be immediately propagated to the model. Please check this test case:
https://github.com/Qiskit/qiskit-experiments/blob/777a59950efdc1c2f718e91f263762f5c6513cdd/test/curve_analysis/test_curve_analysis_base_class.py#L245-L279

nkanazawa1989 commented on test/curve_analysis/fit_options.py:
```diff
@@ -0,0 +1,231 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test for fit options."""
```
oops good catch! thanks.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_data.py:
```diff
@@ -47,6 +48,20 @@ class SeriesDef:
+        signature = list(inspect.signature(self.fit_func).parameters.keys())
+        fitparams = signature[1:]
+
+        # Note that this dataclass is frozen
```
I'm happy to remove but you cannot directly initialize values in post init if the dataclass is frozen. We always need to use setatter.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,270 @@
+    The parameter :math:`\vec{p}_i = \theta_i \cup \Theta_{\rm fix}` is a union of the
+    fit parameter for the function :math:`f_i` and the fixed parameters :math:`\Theta_{\rm fix}`.
+    The composite function :math:`F` consists of multiple fit functions :math:`f_i`
+    taking independent data points :math:`x_i` with partly shared fit parameters :math:`\theta_i`,
```
Done in https://github.com/Qiskit/qiskit-experiments/pull/726/commits/c327fa5f58c74f0017964b31e01828d9d4b9d4d4

nkanazawa1989 commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,270 @@
+
+    This data represents the location where the function with index ``i``
+    is returned and where the x values :math:`x_i` comes from.
+    One must set this data indices before calling the composite fit function.
```
Done in https://github.com/Qiskit/qiskit-experiments/pull/726/commits/c327fa5f58c74f0017964b31e01828d9d4b9d4d4

nkanazawa1989 commented:
Thanks @eggerdj . Indeed I'm thinking of `FitModel.from_config(config: List[SeriesDef])` in the follow-up. Originally this is the PR just to add fit model with minimum functionality. Note that `SeriesDef` and `FitModel` are different object. The series def contains the filtering keyword for data extraction and default drawing properties. Thus, series def = fit model + data processor + drawer. I intentionally exclude filter kwargs and draw options to avoid creating super complicated object that can do everything.

eggerdj commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -996,6 +1100,20 @@ def from_config(cls, config: Union[AnalysisConfig, Dict]) -> "CurveAnalysis":
 
         return instance
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # Remove fit model. Many subclasses define fit functions as lambda function.
+        # In python lambda function is not picklable so instance's fit model is removed here.
+        # This can be easily reconstructed from the class attribute's fit model.
+        del state["_fit_model"]
+        return state
```
```suggestion
    def __getstate__(self):
        """Remove entries from :code:`self.__dict__` that cannot be pickled.
        
        Instances of ``CurveAnalysis`` must be serializable. However, the fit model they contain, i.e. 
        `self._fit_model` has fit functions which are typically defined as lambda functions. These 
        lambda functions are not serializable. Therefore, to make them serializable with pickle we 
        remove them from the state dictionary. Deserialization reconstructs the fit  model from 
        the class fit model and the fixed parameters in the options (see ``__setstate__``).
        """
        state = self.__dict__.copy()
        del state["_fit_model"]
        return state
```

eggerdj commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -996,6 +1100,20 @@ def from_config(cls, config: Union[AnalysisConfig, Dict]) -> "CurveAnalysis":
+        del state["_fit_model"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        # Reconstruct fit model from the class attribute.
```
```suggestion
    def __setstate__(self, state):
        """Reconstruct fit model from the class attribute."""
        self.__dict__.update(state)
```

eggerdj commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -996,6 +1118,20 @@ def from_config(cls, config: Union[AnalysisConfig, Dict]) -> "CurveAnalysis":
 
         return instance
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # Remove fit model. Many subclasses define fit functions as lambda function.
+        # In python lambda function is not picklable so instance's fit model is removed here.
+        # This can be easily reconstructed from the class attribute's fit model.
+        del state["_fit_model"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        # Reconstruct fit model from the class attribute.
+        self._fit_model = self._cls_fit_model.copy()
+        self._fit_model.bind_parameters(**self.options.fixed_parameters)
```
Okay. Understood. This is fine. I made suggestion in the docs to better reflect what is going on an why it is happening.

eggerdj commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,339 @@
+    the corresponding outcome is :math:`Y =\vec{y}_0 \oplus \vec{y}_1 \oplus ...`.
+
+    In the Qiskit Experiments, these data sources are represented by
+    a single flat array ``vec[k]``, rather than a jugged array ``mat[i, j]``.
```
```suggestion
    a single 1D array ``vec[k]``, rather than a 2D array ``mat[i, j]``.
```
I don't know what `jugged` means in this context.

nkanazawa1989 replied:
It may take different element size for every `i`. Then this is not a clean 2D array of NxM.
https://en.wikipedia.org/wiki/Jagged_array

eggerdj commented on qiskit_experiments/curve_analysis/fit_models.py:
```diff
@@ -0,0 +1,270 @@
+    ):
+        super().__init__(fit_functions, signatures, fit_models)
+
+        # This attribute is set by users or another function that calls this model.
```
Ok.

eggerdj reviewed (commented):
I am pretty much fine with this PR my comments are only minor. @yaelbh can you take a second look?

zlatko-minev commented:
The `lmfit` package (Non-Linear Least-Squares Minimization and Curve-Fitting for Python) is pretty powerful with a lot of out of the box features and i've used with good success in the past for fitting qubit experiments on 1D and 2D real and complex data.

It has the `Parameter` and `Parameters` classes that can be also populated by a guess function that can be overwritten by user for iniitial guess. It tracks the convergence etc pretty nicely and ultimately make model fitting and composition of models possible with a decently nice user interface. 

There's a lot of examples here https://lmfit.github.io/lmfit-py/examples/index.html

nkanazawa1989 commented:
Thanks @zlatko-minev I did detailed investigation for `lmfit`. Looks like this is a great tool and we can offload most of functionalities to the fitter there. However there is a blocker in multi-objective optimization (it's called [multiple data sets](https://lmfit.github.io/lmfit-py/examples/example_fit_multi_datasets.html#sphx-glr-examples-example-fit-multi-datasets-py) in the lmfit example). 

This is quite important in our experiment library (e.g. Ramsey XY, fine amplitude, HEAT, CR Hamiltonian tomography, IRB, etc...) but the model interface for this is not supported in the library.
> TODO: this could/should be using the Model interface / built-in models!

For example, in the fine amp example,
```python
from lmfit import Model, CompositeModel
import numpy as np
import matplotlib.pyplot as plt

def spam_cal(x, amp, base):
    return base + 0.5 * amp * (2 * x - 1)

def ping_pong(x, amp, d_theta, phase_offset, base, angle_per_gate):
    return 0.5 * amp * np.cos((d_theta + angle_per_gate) * x - phase_offset) + base

# x values for each model
x1 = np.array([0, 1])
x2 = np.arange(0, 15)

# some parameters
amp = 0.9
d_theta = 0.01
phase_offset = np.pi/2
base = 0.48
angle_per_gate = np.pi

# simulated y data
data1 = spam_cal(x1, amp, base)
data2 = ping_pong(x2, amp, d_theta, phase_offset, base, angle_per_gate)

# standalone model
# How to combine them? Currently lmfit doesn't provide interface to combine.
# Note that this is not CompositeModel where it takes the same x-value and different parameters.
m1 = Model(spam_cal)
m2 = Model(ping_pong)

# so we need to dynamically generate function like below
def composite_func(x, amp, base, d_theta, phase_offset, angle_per_gate, separator):
    # giving these as kwargs is more efficient but lmfit Model parses signature for the model
    # **kwargs doesn't provide any explicit parameter name
    params = {
        "amp": amp,
        "base": base, 
        "d_theta": d_theta, 
        "phase_offset": phase_offset, 
        "angle_per_gate": angle_per_gate,
    } 
    y = []
    for mi, xi in zip((m1, m2), np.split(x, separator)):
        y.append(mi.eval(x=xi, **{p: params[p] for p in mi.param_names}))
    return np.concatenate(y, 0)

# this is what we need
composite_model = Model(composite_func)
```

Considering this, I concluded we cannot employ lmfit library until it supports multi-objective model. However, I still think this is great library because it can provide more statistical information on the fitting, and it can hide uncertainties package (unumpy functions) from the fit model. I'll continue to watch this library.

zlatko-minev commented:
That's interesting. I definitely used it for Ramsey XY fitting and multiple data sets before. You just concatenate the data from the different sets into a large one dimensional array, my experience that usually works pretty well.

nkanazawa1989 commented:
Yes, their fitter (`minimize`) itself supports multi-objective with concatenated array so we need to use the fitter directly. However this doesn't provide any gain over what we have now, i.e. just replacement of the fitter from `spicy.curve_fit` from `lmfit.minimize`. The `Model` interface they implemented (and I'm going to implement with this PR) is nice object to cleanup the data structure, so perhaps it's worth implementing `MultiModel` class by ourselves. 

Another approach would be merging this PR as-is, and write `LmfitCurveAnalysis` and gradually deprecate `CurveAnalysis` to migrate.

nkanazawa1989 commented:
Perhaps concatenated array that you mention is a technique to use `Model` object, i.e. combining two fit functions into a single function with concatenated array. The trickily part is, we allows sub-models to have different signature, for example, 
```
F(x, a, b, c) = F1(x1, a, b) \oplus F2(x2, b, c) where x = x1 \oplus x2 (concatenated)
```
In this case we need to dynamically generate `F` from the signature of `F1` and `F2` without using python `eval`. This is the reason I gave up lmfit.
